src/streams/mbedtls.c

```diff
@@ -303,22 +303,22 @@ static int mbedtls_set_proxy(git_stream *stream, const git_proxy_options *proxy_
 	return git_stream_set_proxy(st->io, proxy_options);
 }
 
-ssize_t mbedtls_stream_write(git_stream *stream, const char *data, size_t len, int flags)
+ssize_t mbedtls_stream_write(git_stream *stream, const char *data, size_t data_len, int flags)
 {
-	size_t read = 0;
+	ssize_t written = 0, len = min(data_len, SSIZE_MAX);
 	mbedtls_stream *st = (mbedtls_stream *) stream;
 
 	GIT_UNUSED(flags);
 
 	do {
-		int error = mbedtls_ssl_write(st->ssl, (const unsigned char *)data + read, len - read);
+		int error = mbedtls_ssl_write(st->ssl, (const unsigned char *)data + written, len - written);
 		if (error <= 0) {
 			return ssl_set_error(st->ssl, error);
 		}
-		read += error;
-	} while (read < len);
+		written += error;
+	} while (written < len);
 
-	return read;
+	return written;
```

Comment:
So instead of writing all bytes, we just write the first `SSIZET_MAX` bytes. This unfortunately breaks callers writing more than `SSIZET_MAX` bytes, as they never check for partially written buffers but for errors, only. E.g.:

```
static int write_chunk(git_stream *io, const char *buffer, size_t len)
{
	git_buf buf = GIT_BUF_INIT;

	/* Chunk header */
	git_buf_printf(&buf, "%" PRIxZ "\r\n", len);

	if (git_buf_oom(&buf))
		return -1;

	if (git_stream_write(io, buf.ptr, buf.size, 0) < 0) {
		git_buf_dispose(&buf);
		return -1;
	}

	git_buf_dispose(&buf);

	/* Chunk body */
	if (len > 0 && git_stream_write(io, buffer, len, 0) < 0)
		return -1;

	/* Chunk footer */
	if (git_stream_write(io, "\r\n", 2, 0) < 0)
		return -1;

	return 0;
}
```

Reply:
Yup.  I have a TODO item to fix these up also.  I could raise this as an issue or something, but it needs a serious audit.  It looks like we have several places where we assume that `write` (maybe `read`, too) will always write everything we ask it to.

I was thinking about swapping in `p_read` with a function that works byte-at-a-time in our tests to flush out all offenders.

But back to this:  if we ever did write all bytes (where the buffer size was greater than `SSIZE_MAX`) then we'd wrap and end up returning an error.  Probably the current behavior is slightly better (thinking we failed when we didn't, rather than writing half the bytes and thinking we succeeded)...  but we're rather hosed either way.

In any case, I'd be willing to go through and make sure that we always read/write in a loop before merging this or after.  The risk of saying "oh we'll do it later" is, of course, that we never actually do it.

Reply:
Just let me know if you've got the time. Otherwise I can also rip this out and handle it myself, as I should have some time to do so this week

Reply:
I likely won't be able to get to this until the weekend so you're welcome to if you'd like.

Reply:
Both `p_read` and `p_write` are fine, as we already loop around read(3) and write(3)

Reply:
See #4953